### PR TITLE
feat(knowbase): add entity on articles

### PR DESCRIPTION
**Pull request by Rom1-B**
Addition of the entity on the articles of the KB, to have a partitioning like with the other objects of GLPI.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !25522 + !24105

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
Aren't entities for articles already defined in the "criteria" tab ?

**Comment by Rom1-B** [comment]
> Aren't entities for articles already defined in the "criteria" tab ?

Yes, for distribution, but until now, the article did not belong to any entity, so an admin had access to all existing articles, even if his profile is limited to a single sub-entity.

**Comment by cconard96** [comment]
So, basically the article itself will have a default or management entity that can then be shared with other entities?

It will only be editable from its own entity, but viewable in any it is shared with?

**Comment by Rom1-B** [comment]
The goal is that the entities on the items work as elsewhere (e.g. computers).

If the item is in a sub-entity, it is only visible in this entity and its children.

In the targets by entities, only the entity of the article and its children are proposed.

If the article is in a root entity, this is exactly how it works now.

File: install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php [hunk 1]
```diff
@@ -37,3 +37,33 @@
...
 $ADDTODISPLAYPREF[KnowbaseItem::class] = [79, 131, 13];
+
+$table = 'glpi_knowbaseitems';
+
+$fkey_to_add = 'entities_id';
+if (!$DB->fieldExists($table, $fkey_to_add)) {
+    $migration->addField(
+        $table,
+        $fkey_to_add,
+        "int {$default_key_sign} NOT NULL DEFAULT '0'",
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php:48]: You can use the corresponding format helper.
```suggestion
        'fkey',
```

File: install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php [hunk 2]
```diff
@@ -37,3 +37,33 @@
...
+        [
+            'value' => 0,
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php:50]: No need to define it as it is already defined by `fkey` type.
```suggestion
```

File: install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php [hunk 3]
```diff
@@ -37,3 +37,33 @@
...
+}
+
+$fkey_to_add = 'is_recursive';
+if (!$DB->fieldExists($table, $fkey_to_add)) {
+    $migration->addField(
+        $table,
+        $fkey_to_add,
+        "tinyint NOT NULL DEFAULT '0'",
+        [
+            'value' => 1,
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php:64]: `value` is for field default value. If you want to update existing value, you have to use `update` option.

```suggestion
            'update' => 1,
```

File: install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php [hunk 4]
```diff
@@ -37,3 +37,33 @@
...
+    );
+    $migration->addKey($table, $fkey_to_add);
+}
+
+$fkey_to_add = 'is_recursive';
+if (!$DB->fieldExists($table, $fkey_to_add)) {
+    $migration->addField(
+        $table,
+        $fkey_to_add,
+        "tinyint NOT NULL DEFAULT '0'",
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php:62]: You can use the corresponding format helper.
```suggestion
        'bool',
```

File: install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php [hunk 5]
```diff
@@ -37,3 +37,33 @@
...
+        "int {$default_key_sign} NOT NULL DEFAULT '0'",
+        [
+            'value' => 0,
+            'after' => 'id'
+        ]
+    );
+    $migration->addKey($table, $fkey_to_add);
+}
+
+$fkey_to_add = 'is_recursive';
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/knowbaseitem.php:57]: This is not a fkey, variable name should be changed.
```suggestion
$field_to_add = 'is_recursive';
```

File: src/CommonDBVisible.php [hunk 6]
```diff
@@ -434,8 +434,10 @@ public function showVisibility()
     protected function getShowVisibilityDropdownParams()
     {
         return [
-            'type'  => '__VALUE__',
-            'right' => strtolower($this::getType()) . '_public'
+            'type'          => '__VALUE__',
+            'right'         => strtolower($this::getType()) . '_public',
+            'entity'        => $this->fields['entities_id'],
+            'is_recursive'  => $this->fields['is_recursive'],
```
Comment (cedric-anne) [inline src/CommonDBVisible.php:440]: `getShowVisibilityDropdownParams()` is also used on plugins. New parameters should only be added if there is a corresponding field.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**